thorlcr/master/thdemonserver.cpp

```diff
@@ -165,7 +179,47 @@ class DeMonServer : public CSimpleInterface, implements IDeMonServer
             }
         }
     }
-    void reportGraph(CGraphBase *graph, bool finished, bool success, unsigned startTime, unsigned __int64 startTimeStamp)
+    void reportCompletedGraphs()
+    {
+        for (auto graph: finishedGraphs)
+        {
+            try
+            {
+                IConstWorkUnit &currentWU = graph->queryJob().queryWorkUnit();
+                const char *graphName = graph->queryJob().queryGraphName();
+                unsigned wfid = graph->queryJob().getWfid();
+                Owned<IWUGraphStats> stats = currentWU.updateStats(graphName, SCTthor, queryStatisticsComponentName(), wfid, graph->queryGraphId(), false);
+                statisticsCache->mergeInto(wfid, graph->queryGraphId(), stats->queryStatsBuilder(), true);
+                Owned<IWorkUnit> wu = &currentWU.lock();
+                queryServerStatus().commitProperties();
+            }
+            catch (IException *e)
+            {
+                StringBuffer s;
+                LOG(MCwarning, thorJob, "Failed to update graph stats information: %s", e->errorMessage(s).str());
+                e->Release();
+            }
+        }
+        finishedGraphs.clear();
+    }
+    void gatherGraphStats(CGraphBase *graph, bool finished)
+    {
+        try
+        {
+            unsigned wfid = graph->queryJob().getWfid();
+            if (finished)
+                finishedGraphs.push_back(LINK(graph));
+            Owned<IStatisticGatherer> gatherer = createStatsAggregatorGather(statisticsCache, wfid, graph->queryGraphId());
```

Comment:
This may or may not answer your question(!)

The overall design should be included as a comment - especially because there are potential problems which need to be addressed and documented.

I came up with two possible implementations:
1) Simple cache
a) Have a cache of mapping (scope)->stats
b) when gather stats for subgraph also publish to cache/aggregator by replacing the existing values, and marking this and all parents as dirty.
c) to publish for a graph.etc walk all levels of the cache.  if dirty recalculate by summing childman +publishing.

Problems:
You need to preload the stats for all other children.  E.g. other graphs in the same workflow, other workflows in the same workunit. (And subgraphs for a graph if we improve restart).  If they are all held in the same place (wokunit stats) then that should be fairly efficient.

2) A delta cache (I think the solution Shamser has used)

b) when publish replace values and recursively apply a delta to the parent (and mark as dirty)
c) When ready to publish, publish all dirty stats.

Problems:
Either need to apply a delta when values are published, or need to load previous values (e.g. for workflow) from the workunit.
If subgraphs are rerun (e.g., because a job has failed) you will need a mechanism to not include the previous values, otherwise the aggregations will include both runs.

As far as implementation goes I would explore two options.

i) A hashtable mapping (scope string)->CRuntimeCollection.  Probably with a parent pointer as well.  Use a StatisticsMapping to define which stats are aggregated
ii) Use a CStatisticsCollection.  This already supports nested scopes and arbitrary stats.  It is likely to need some extension (e.g, tracking whether dirty).